File: src/services/payment.ts
import Stripe from 'stripe';
import { Order } from '../models/Order';
import { Address } from '../models/Address';

const stripe = new Stripe(process.env.STRIPE_SECRET_KEY!, {
  apiVersion: '2023-10-16',
});

export const createPaymentIntent = async (order: Order, shippingAddress: Address, billingAddress: Address) => {
  try {
    // Create payment intent
    const paymentIntent = await stripe.paymentIntents.create({
      amount: Math.round(order.total_amount * 100), // Convert to cents
      currency: 'usd',
      metadata: {
        order_id: order.id,
        user_id: order.user_id,
      },
    });

    // Create shipping address in Stripe
    const shippingAddressId = await stripe.customers.create({
      email: order.user.email,
      name: shippingAddress.full_name,
      address: {
        line1: shippingAddress.address_line1,
        line2: shippingAddress.address_line2,
        city: shippingAddress.city,
        state: shippingAddress.state,
        postal_code: shippingAddress.postal_code,
        country: shippingAddress.country,
      },
      phone: shippingAddress.phone_number,
    });

    // Create billing address in Stripe
    const billingAddressId = await stripe.customers.create({
      email: order.user.email,
      name: billingAddress.full_name,
      address: {
        line1: billingAddress.address_line1,
        line2: billingAddress.address_line2,
        city: billingAddress.city,
        state: billingAddress.state,
        postal_code: billingAddress.postal_code,
        country: billingAddress.country,
      },
      phone: billingAddress.phone_number,
    });

    return {
      clientSecret: paymentIntent.client_secret,
      shippingAddressId: shippingAddressId.id,
      billingAddressId: billingAddressId.id,
    };
  } catch (error) {
    console.error('Error creating payment intent:', error);
    throw new Error('Error creating payment intent.');
  }
};

export const handleWebhook = async (event: Stripe.Event) => {
  try {
    switch (event.type) {
      case 'payment_intent.succeeded':
        const paymentIntent = event.data.object as Stripe.PaymentIntent;
        const orderId = paymentIntent.metadata.order_id;

        // Update order status
        await Order.update(
          { status: 'processing' },
          { where: { id: orderId } }
        );
        break;

      case 'payment_intent.payment_failed':
        const failedPaymentIntent = event.data.object as Stripe.PaymentIntent;
        const failedOrderId = failedPaymentIntent.metadata.order_id;

        // Update order status
        await Order.update(
          { status: 'cancelled' },
          { where: { id: failedOrderId } }
        );
        break;

      default:
        console.log(`Unhandled event type ${event.type}`);
    }
  } catch (error) {
    console.error('Error handling webhook:', error);
    throw new Error('Error handling webhook.');
  }
};

export const createRefund = async (order: Order) => {
  try {
    const refund = await stripe.refunds.create({
      payment_intent: order.payment_intent_id,
      reason: 'requested_by_customer',
    });

    // Update order status
    await Order.update(
      { status: 'refunded' },
      { where: { id: order.id } }
    );

    return refund;
  } catch (error) {
    console.error('Error creating refund:', error);
    throw new Error('Error creating refund.');
  }
}; 

File: app/services/payment.py
import stripe
from app.core.config import settings
from typing import Dict, Any

stripe.api_key = settings.STRIPE_SECRET_KEY

async def create_payment_intent(
    amount: float,
    currency: str = "usd",
    metadata: Dict[str, Any] = None
) -> stripe.PaymentIntent:
    """
    Create a Stripe payment intent.
    """
    try:
        payment_intent = stripe.PaymentIntent.create(
            amount=int(amount * 100),  # Convert to cents
            currency=currency,
            metadata=metadata,
            automatic_payment_methods={
                "enabled": True,
            },
        )
        return payment_intent
    except stripe.error.StripeError as e:
        raise Exception(f"Failed to create payment intent: {str(e)}")

async def confirm_payment_intent(
    payment_intent_id: str,
    payment_method_id: str
) -> stripe.PaymentIntent:
    """
    Confirm a Stripe payment intent.
    """
    try:
        payment_intent = stripe.PaymentIntent.confirm(
            payment_intent_id,
            payment_method=payment_method_id,
        )
        return payment_intent
    except stripe.error.StripeError as e:
        raise Exception(f"Failed to confirm payment: {str(e)}")

async def refund_payment(
    payment_intent_id: str,
    amount: float = None
) -> stripe.Refund:
    """
    Refund a payment.
    """
    try:
        refund_params = {
            "payment_intent": payment_intent_id,
        }
        if amount:
            refund_params["amount"] = int(amount * 100)  # Convert to cents
            
        refund = stripe.Refund.create(**refund_params)
        return refund
    except stripe.error.StripeError as e:
        raise Exception(f"Failed to process refund: {str(e)}")

async def get_payment_intent(
    payment_intent_id: str
) -> stripe.PaymentIntent:
    """
    Get a payment intent by ID.
    """
    try:
        payment_intent = stripe.PaymentIntent.retrieve(payment_intent_id)
        return payment_intent
    except stripe.error.StripeError as e:
        raise Exception(f"Failed to get payment intent: {str(e)}")

async def create_payment_method(
    card_details: Dict[str, Any]
) -> stripe.PaymentMethod:
    """
    Create a payment method from card details.
    """
    try:
        payment_method = stripe.PaymentMethod.create(
            type="card",
            card={
                "number": card_details["number"],
                "exp_month": card_details["exp_month"],
                "exp_year": card_details["exp_year"],
                "cvc": card_details["cvc"],
                "billing_details": card_details.get("billing_details", {}),
            },
        )
        return payment_method
    except stripe.error.StripeError as e:
        raise Exception(f"Failed to create payment method: {str(e)}") 